### ISSUE TITLE
Better FID calculation.

### ISSUE THREAD

**Issue by Adversarian** [lucidrains/denoising-diffusion-pytorch#224]
Hi,
while reading (and using) the code I noticed that the current FID calculation is dependent on the training batch size. That is, the total number of real samples extracted for FID calculation is at most equal to training batch size ([line 972 here](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py)) while it is empirically shown that for FID to be meaningful as a metric, the number of samples should be in the ballpark of ~50k which is an impossible batch size for a diffusion model no matter how much computation power you may possess.

I propose that if FID calculation is requested from the `Trainer` (by setting `calculate_fid` to `True`), we also take the number of samples as a keyword arg (say `num_samples_for_fid` with a default value to set to `50000`). Then we precalculate the mean and variance of the `num_samples_for_fid` real samples from the dataset and either save them to disk to free up memory temporarily until the next FID calculation milestone or just keep them loaded as an attribute of the class (`self.m1` and `self.s1` for instance). 

Finally, at `save_and_sample_every` steps we calculate an unbiased FID using `self.m1` and `self.s1` and the generated samples. Note that we only need to save the feature of samples obtained by running them through the InceptionV3 network and stack them on top of each other so that there are no memory issues from keeping the original images.

Granted, this takes significantly longer to calculate specially if DDIM sampling is disabled but it should give a more meaningful FID which can also potentially serve as a means to save only the `best` and `latest` models instead of saving a checkpoint at every milestone.

If this sounds compelling to you I'd be more than happy to submit a PR when I find some time as I have already monkey-patched these changes in my clone of the repo.

**Comment by lhaippp** [comment]
Hi, 
I think it's a good idea, as we can get a more meaningful FID score.
Meanwhile, as the checkpoints are a little bit large (800M), saving best and latest models seem to be reasonable.
If you submit a PR, I'm willing to do a Code Review @Adversarian 

How do you think? @lucidrains

**Comment by lucidrains** [comment]
yea, this is a good idea

i can take this on and complete before end of month. i may move all the eval logic into a separate file too, as i don't want to obscure from the main research code

**Comment by Adversarian** [comment]
> yea, this is a good idea
> 
> i can take this on and complete before end of month. i may move all the eval logic into a separate file too, with some `FidEvaluation` module that takes care of the precomputed statistics, caching, etc. i don't want to obscure the main research code with a bunch of irrelevant engineering

Incidentally this is what I did as well because I felt like having 1000+ lines in a single file can be a little off-putting to go through and maintain. If you're going to do it, I don't want wanna step on any toes but I'd be happy to do it if you don't have time. It'll be a great engineering exercise for me.

**Comment by lucidrains** [comment]
oh yeah sure! I missed the last paragraph where you offered to PR

maybe for simplicity's sake, just assume it is single machine
